### ISSUE TITLE
Quick-fix to create a class from its call expression

### ISSUE THREAD

**Issue by nickebbitt** [nicoespeon/hocus-pocus#9]
## Is this request related to a problem? Please describe.

Create a class declaration from its call expression using a quick-fix.

## Describe the solution you'd like

Similarly to the create function feature, it's nice, especially when writing test-first, to be able to write some code that described how you'd like to create a new class with its dependencies that will be injected via the constructor.

An example could be:

```
const aString: string = "test";
const aNumber: number = 1;

new MyClass(aString, aNumber);
```

Then a quick-fix from `MyClass` would generate a class such as:

```
class MyClass {
   constructor(private readonly aString: string, private readonly aNumber: number) {
     this.aString = aString;
     this.aNumber = aNumber;
  }
}
```

## Additional context

Coming from a Java background using the JetBrains IntelliJ (with similar support as WebStorm) it's really easy to just quick-fix to generate the necessary code.

**Comment by nicoespeon** [comment]
Yep @nickebbitt definitely! Would love to have that implemented, I can definitely use it too 👌 

I think we should handle JS and TS kinda separately for this one. In JS we can't use the auto-assignment.

So in JS, it would be:

```js
class MyClass {
   constructor(aString, aNumber) {
     this.aString = aString;
     this.aNumber = aNumber;
  }
}
```

In TS, we can directly use the auto-assignment feature and start with:

```ts
class MyClass {
   constructor(public readonly aString, public readonly aNumber) {}
}
```

Finally, if we solve #3 (get the inferred types in TS) we can probably produce:

```ts
class MyClass {
   constructor(public readonly aString: string, public readonly aNumber: number) {}
}
```

I'd be fine if we implement this progressively, stage by stage.

**Comment by nickebbitt** [comment]
Awesome, I'll start to familiarise myself with the project and vscode extensions with a view get something working for the TS impl as a first step.

**Comment by nickebbitt** [comment]
Hey @nicoespeon - I've managed to find some time to make some initial progress and have started to get my head around the AST and running the extension in debug mode.

I've been using the `create-function` impl as a template and reviewing the contexts under which the auto-fix should be available and feel generally they are quite similar.

One scenario that's interesting is with `await`.

```
async function myFunc() {
  const code = await new MyClass().doSomething();
}
```

The create class auto-fix can give us:
```
class MyClass {
  // Implement
}
```
which can be followed up with a "create method" auto-fix to give us:
```
class MyClass {
  doSomething() {
    throw new Error("Method not implemented.");
  }
}
```
Unfortunately the create method auto-fix is limited in that it is not async/await aware so an extra manual step is currently necessary, maybe one for the future.

Generally I would likely refactor the calling expression to introduce a variable like:
```
async function myFunc() {
  const myClass = new MyClass();
  const code = await myClass.doSomething();
}
```
which seems quite nice :)

Anyway, I mainly just wanted to check that following the create function impl as a template is a good path to follow? If so, I'll aim to get a draft PR up fairly soon to get some early feedback to ensure things look like they are on track to meet the expectations of the project.

**Comment by nickebbitt** [comment]
@nicoespeon what are your thoughts on how we implement auto-assignment in the constructor? 

The example you provided specifies the args as `public readonly`:

```
class MyClass {
   constructor(public readonly aString, public readonly aNumber) {}
}
```

My instinct would be specify them as `private`, favouring encapsulation as the default:

```
class MyClass {
   constructor(private readonly aString, private readonly aNumber) {}
}
```

It's possible this is my OO bias coming through from the Java ecosystem though so happy to go with whatever is the JS/TS idiom.

**Comment by nicoespeon** [comment]
@nickebbitt yes, it should be private by default, it was my mistake 👍 

Also, good job in following the template. With more duplication I might find the correct abstractions later. For the moment, it works fine 👌

**Comment by nicoespeon** [comment]
@nickebbitt first draft has been released under v1.3.0.

Thank you again 👍 

Feel free to iterate with further PRs. In the meantime, we're working on getting "Create Function" work with types: #3 
When it's done, we'll be able to do the same for TS classes!
